src/core/mount.c

```diff
@@ -84,15 +84,23 @@ static MountParameters* get_mount_parameters(Mount *m) {
         return get_mount_parameters_fragment(m);
 }
 
-static bool mount_is_network(const MountParameters *p) {
-        assert(p);
+static int mount_is_network(Mount *m) {
+        MountParameters *p;
 
-        if (fstab_test_option(p->options, "_netdev\0"))
-                return true;
+        p = get_mount_parameters(m);
+        if (!p)
+                return -1;
 
         if (p->fstype && fstype_is_network(p->fstype))
                 return true;
 
+        if (m->from_proc_self_mountinfo &&
+            UNIT(m)->manager->last_libmount_event_type != MNT_MONITOR_TYPE_USERSPACE)
```

Comment:
Hmm... `mount_is_network()` is called in the several paths:
- mount_load() -> mount_add_extras() -> mount_add_non_exec_dependencies()
- mount_setup_unit() -> mount_setup_{new,existing}_unit() -> mount_add_non_exec_dependencies()

What happens the following?
1. the manager receives a userspace event which is triggered by a path e.g. /pathA,
2. the manager receives a kernel event triggered by another path e.g. /pathB,
3. then, the mount unit for /pathA gets `MOUNT_PROC_JUST_CHANGED` flag...
4. and subsequent call of `mount_is_network()` for /pathA returns negative...

Does this patch really work as expected??

Anyway, please add a testcase.

Reply:
I've just noticed that even with MNT_FS_MERGED, this patch never fix the issue. Consider again that mount command updates /run/mount/utab file only when mount options contain those for user-mode applications such as _netdev. For such mount entries whose mount options not containing these for user-mode applications, MNT_FS_MERGED is never set. No matter how long we wait, we cannot determine whether a given mount entry is local one or network one for such mount entries. I wish libmount updates utab at all mount.